// ngx-translate-multi-http-loader/esm2015/lib/multi-http-loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { forkJoin, of } from "rxjs";
import { catchError, map } from "rxjs/operators";
import merge from 'deepmerge';
/**
 * @record
 */
export function ITranslationResource() { }
function ITranslationResource_tsickle_Closure_declarations() {
    /** @type {?} */
    ITranslationResource.prototype.prefix;
    /** @type {?} */
    ITranslationResource.prototype.suffix;
}
export class MultiTranslateHttpLoader {
    /**
     * @param {?} http
     * @param {?} resources
     */
    constructor(http, resources) {
        this.http = http;
        this.resources = resources;
    }
    /**
     * @param {?} lang
     * @return {?}
     */
    getTranslation(lang) {
        const /** @type {?} */ requests = this.resources.map(resource => {
            const /** @type {?} */ path = resource.prefix + lang + resource.suffix;
            return this.http.get(path).pipe(catchError(res => {
                console.error("Could not find translation file:", path);
                return of({});
            }));
        });
        return forkJoin(requests).pipe(map(response => merge.all(response)));
    }
}
function MultiTranslateHttpLoader_tsickle_Closure_declarations() {
    /** @type {?} */
    MultiTranslateHttpLoader.prototype.http;
    /** @type {?} */
    MultiTranslateHttpLoader.prototype.resources;
}
