// getsocial-cordova-sdk/plugin/im.getsocial.sdk.cordova/www/plugin/InvitePlugin.js
/*
 *      Copyright 2015-2016 GetSocial B.V.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

  var Plugin = require('./Plugin')

  var NATIVE_BRIDGE_NAME = "GetSocialPluginEvents";

  var InvitePlugin = function() {
  }

  InvitePlugin.prototype.isAvailableCallbackID = function() {
    console.log(this.callbackId);
    if(this.callbackId == null)
      return false

    return true
  }

  InvitePlugin.prototype.completeCallback = function(s, list) {
    if(this.isAvailableCallbackID())
      cordova.exec(null, null, NATIVE_BRIDGE_NAME, "completeCallback", [this.callbackId, s, list]);
  }
  InvitePlugin.prototype.cancelCallback = function() {
    if(this.isAvailableCallbackID())
      cordova.exec(null, null, NATIVE_BRIDGE_NAME, "cancelCallback", [this.callbackId]);
  }
  InvitePlugin.prototype.errorCallback = function() {
    if(this.isAvailableCallbackID())
      cordova.exec(null, null, NATIVE_BRIDGE_NAME, "errorCallback", [this.callbackId]);
  }
  InvitePlugin.prototype.clickedCallback = function() {
    if(this.isAvailableCallbackID())
      cordova.exec(null, null, NATIVE_BRIDGE_NAME, "clickedCallback", [this.callbackId]);
  }


  InvitePlugin.prototype.inviteFriends = function(subject, text, referralDataUrl, image, onSuccess, onCancel, onFailure) {}

  InvitePlugin.prototype.__proto__ = Plugin.prototype;

  module.exports = InvitePlugin;
