// getsocial-cordova-sdk/plugin/im.getsocial.sdk.cordova/www/entity/LeaderboardMetaData.js
/*
 *      Copyright 2015-2016 GetSocial B.V.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

var LeaderboardMetaData = function(serializedObject) {
    this.leaderboardId = "";
    this.name = "";
    this.units = "";
    this.direction = 0;
    this.format = "";
    this.published = false;
    this.deserialize(serializedObject);
}

LeaderboardMetaData.prototype.deserialize = function(object) {
    if(object == null)
        return null;
    this.leaderboardId = object.leaderboardId;
    this.name = object.name;
    this.units = object.units;
    this.direction = object.direction;
    this.format = object.format;
    this.published = object.published;
}

LeaderboardMetaData.prototype.getLeaderboardId = function() {
    return this.leaderboardId;
}

LeaderboardMetaData.prototype.getName = function() {
    return this.name;
}

LeaderboardMetaData.prototype.getUnits = function() {
   return this.units;
}

LeaderboardMetaData.prototype.getDirection = function() {
   return this.direction;
}

LeaderboardMetaData.prototype.getFormat = function() {
   return this.format;
}

LeaderboardMetaData.prototype.isPublished = function() {
   return this.published;
}

module.exports = LeaderboardMetaData;
